main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -114,30 +114,37 @@ object Kinder {
    */
   private def visitStruct(struct0: ResolvedAst.Declaration.Struct, taenv: Map[Symbol.TypeAliasSym, KindedAst.TypeAlias], root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Struct, KindError] = struct0 match {
     case ResolvedAst.Declaration.Struct(doc, ann, mod, sym, tparams0, fields0, loc) =>
-      // if not annotated tparams are assumed to be Star except the last one which is Eff
-      val tparams1 = tparams0
-      val kenv0 = getKindEnvFromTypeParams(tparams1)
-      val kenvVal = kenv0 + (tparams1.last.sym -> Kind.Eff)
-
+      // In the case in which the user doesn't supply any type params,
+      // the parser will have already notified the user of this error
+      // The recovery step here is to simply add a single type param that is never used
+      val tparams1 = if (tparams0.isEmpty) {
+        val regionTparam = ResolvedAst.TypeParam.Unkinded(Name.Ident("$rc", loc), Symbol.freshUnkindedTypeVarSym(Ast.VarText.Absent, isRegion = false, loc), loc)
+        List(regionTparam)
+      } else {
+        tparams0
+      }
+      // if not annotated, tparams are assumed to be Star except the last one which is Eff
+      val tparams2 = tparams1.init.map(makeKinded(_, Kind.Star)) :+ makeKinded(tparams1.last, Kind.Eff)
```

Comment:
This is closer, but not quite what I had in mind.

To be consistent with other declarations, the pattern should be:
1. Infer kinds (creating a kind environment)
2. Use that kind environment to update the declaration

So here we should start by building a kind environment from all except the last parameter by using `getKindEnvFromTypeParams`. Then we visit the last parameter with a new function `getKindEnvFromRegion` or something, which defaults to `Eff` rather than `Star`.

So now we have 2 Kind environments. We merge the two and carry on by visiting the type params (rather than "manually" transforming them) and calling the appropriate `visitX` on all the other stuff.

Reply:
This logic should somewhat resemble what goes on with `RestrictableEnum` since that also has a parameter (`index`) which must have a particular kind.

Reply:
Oh, I think I understand a bit better. Does what I have now make more sense?

Reply:
Yes, this looks good.